Fix cgroup hugetlb size prefix for kB

odinuge opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
The hugetlb cgroup control files (introduced here in 2012:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=abb8206cb0773)
use "KB" and not "kB"
(https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/mm/hugetlb_cgroup.c?h=v5.0#n349).

The behavior in the kernel has not changed since the introduction, and the current code using "kB" will therefore fail on devices with small amounts of ram (see #77169) running a kernel with config flag CONFIG_HUGETLBFS=y

As seen from the code in "mem_fmt" inside hugetlb_cgroup.c, only "KB",
"MB" and "GB" are used, so the others may be removed as well.

Here is a real world example of the files inside the
"/sys/kernel/mm/hugepages/" directory:

- "hugepages-64kB"
- "hugepages-2048kB"
- "hugepages-32768kB"
- "hugepages-1048576kB"

And the corresponding cgroup files:
- "hugetlb.64KB._____"
- "hugetlb.2MB._____"
- "hugetlb.32MB._____"
- "hugetlb.1GB._____"

Noticed this when tinkering around with kubernetes on a Raspberry PI (1GB ram) running Arch Linux (`Linux k8s-003 5.1.5-1-ARCH #1 SMP Sat May 25 13:23:49 MDT 2019 aarch64 GNU/Linux`)



**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #77169

**Special notes for your reviewer**:

/priority important-soon

~`WIP` beacuse it depends on this upstream fix in runc:
https://github.com/opencontainers/runc/pull/2065~


**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Fix kubelet errors in AArch64 with huge page sizes smaller than 1MiB
```

odinuge committed:
Fix cgroup hugetlb size prefix for kB

Use the exported list from runc that uses "KB" and not "kB".

This issue breaks kubelet on AArch64 (arm 64).

var HugePageSizeUnitList = []string{"B", "KB", "MB", "GB", "TB", "PB"}

The hugetlb cgroup control files (introduced here in 2012:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=abb8206cb0773)
use "KB" and not "kB"
(https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/mm/hugetlb_cgroup.c?h=v5.0#n349).

The behavior in the kernel has not changed since the introduction, and
the current code using "kB" will therefore fail on devices with huge
pages smaller than 1MiB. This is the case for AArch64.

As seen from the code in "mem_fmt" inside hugetlb_cgroup.c, only "KB",
"MB" and "GB" are used, so the others may be removed as well.

Here is a real world example of the files inside the
"/sys/kernel/mm/hugepages/" directory:
- "hugepages-64kB"
- "hugepages-2048kB"
- "hugepages-32768kB"
- "hugepages-1048576kB"

And the corresponding cgroup files:
- "hugetlb.64KB._____"
- "hugetlb.2MB._____"
- "hugetlb.32MB._____"
- "hugetlb.1GB._____"

Signed-off-by: Odin Ugedal <odin@ugedal.com>

odinuge commented:
CLA should be good to go now.

mattjmcnaughton reviewed (commented):
Thanks for your work on this issue @odinuge :)

I think the largest question I'm trying to wrap my head around is whether this feature ever worked, or if it has been broken since it was first introduced?

If the former, will this change cause issues for those for whom it was working before?

Additionally, thoughts on how hard it would be to add a unit/integration test to assert this behavior is working?

Lmk when the change on which this depends is merged, and I'll mark this as ok-to-test.

odinuge commented:
Hi @mattjmcnaughton,

I have twisted my head around this quite a few times myself. HugePages smaller than 1MB doesn't make that much sense after all... And the code is kinda strange, as some stuff is done inside runc, and more or less the same is done inside k8s.

After a bit more research I think I found a relavant patch (doing changes inside architecture code of arm64 only), https://lkml.org/lkml/2018/10/23/143. The line `add_huge_page_size(PMD_SIZE * CONT_PMDS)` with the config of the kernel I use, ends up as `PMD_SIZE = 1UL << 12` and `CONT_PMDS = 1UL << 4`. Resulting in `add_huge_page_size(1UL << 16)`, aka. HugePage size of `64kB`.

So, it looks like this that patch introduced the HugePage size `64kB` in arm64, even tho such small HugePages has been supported in the kernel from day 1.

> If the former, will this change cause issues for those for whom it was working before?

No. Using `kB` instead of `KB` for the cgroup(v1) control files has never worked (the kernel has used `KB` from day 1), but as I said over, i guess no one would have seen this before the patch a referred to. :smile: 

> Additionally, thoughts on how hard it would be to add a unit/integration test to assert this behavior is working?

Yee, I do indeed prefer writing some tests. The file containing the code has ~18% coverage, but the func using it has 0.. :( Guess I can write a simple test ensuring KB is used, but I don't think that would give any value, or, what do you think?


Also, as I said in the runc-PR, this looks like the result of a quick copy-paste of the ByteSizes, meant for making huge numbers more human readable, from the docker/go-units package. https://github.com/docker/go-units/blob/519db1ee28dcc9fd2474ae59fca29a810482bfb1/size.go#L37. And another copy-paste from `runc` into `kubernetes` :smile: 

> Lmk when the change on which this depends is merged, and I'll mark this as ok-to-test.

Will do :smile:

mattjmcnaughton commented:
> Hi @mattjmcnaughton,
> 
> I have twisted my head around this quite a few times myself. HugePages smaller than 1MB doesn't make that much sense after all... And the code is kinda strange, as some stuff is done inside runc, and more or less the same is done inside k8s.
> 
> After a bit more research I think I found a relavant patch (doing changes inside architecture code of arm64 only), https://lkml.org/lkml/2018/10/23/143. The line `add_huge_page_size(PMD_SIZE * CONT_PMDS)` with the config of the kernel I use, ends up as `PMD_SIZE = 1UL << 12` and `CONT_PMDS = 1UL << 4`. Resulting in `add_huge_page_size(1UL << 16)`, aka. HugePage size of `64kB`.
> 
> So, it looks like this that patch introduced the HugePage size `64kB` in arm64, even tho such small HugePages has been supported in the kernel from day 1.
> 
> > If the former, will this change cause issues for those for whom it was working before?
> 
> No. Using `kB` instead of `KB` for the cgroup(v1) control files has never worked (the kernel has used `KB` from day 1), but as I said over, i guess no one would have seen this before the patch I referred to. 
> 
> > Additionally, thoughts on how hard it would be to add a unit/integration test to assert this behavior is working?
> 
> Yee, I do indeed prefer writing some tests. The file containing the code has ~18% coverage, but the func using it has 0.. :( Guess I can write a simple test ensuring KB is used, but I don't think that would give any value, or, what do you think?
> 
> Also, as I said in the runc-PR, this looks like the result of a quick copy-paste of the ByteSizes, meant for making huge numbers more human readable, from the docker/go-units package. https://github.com/docker/go-units/blob/519db1ee28dcc9fd2474ae59fca29a810482bfb1/size.go#L37. And another copy-paste from `runc` into `kubernetes` 
> 
> > Lmk when the change on which this depends is merged, and I'll mark this as ok-to-test.
> 
> Will do

Thanks for your detail here :) Tbh, I don't have a ton of experience with the cgroup manager component of the kubelet, so I'm not positive what exact type of test coverage already exists. At the very least, adding a unit test would be helpful. It would also be interesting to have a integration testing the entire huge page workflow (if that's possible). I'm not sure how difficult that integration test would be to write (and it also may be more appropriate to do it outside of this diff).

odinuge commented:
After some considerations I made the list of units public/expoted in `runc`, and wrote a few tests there instead: https://github.com/opencontainers/runc/pull/2065/files. When/If that change is merged into upstream, we can just that constant instead, and avoid any problems with regression. :smile: 

A integration test for cgroup (all the different parts) handling inside kubelet would be nice, but that is way outside my expertise atm... Without some serious rewrites, doing that in a general way (for testing this particular problem) would be quite hard I imagine. Internally `runc` does: `files, err := ioutil.ReadDir("/sys/kernel/mm/hugepages")`, aka. reads directly from the system. Since only aarch64(armv8 64bits, not 100% sure about arm32) have hugePages smaller than `1MB`, some serous file mocking, and code rewriting, would be required to test it on other architectures.

mattjmcnaughton commented:
> After some considerations I made the list of units public/expoted in `runc`, and wrote a few tests there instead: https://github.com/opencontainers/runc/pull/2065/files. When/If that change is merged into upstream, we can just that constant instead, and avoid any problems with regression. 
> 
> A integration test for cgroup (all the different parts) handling inside kubelet would be nice, but that is way outside my expertise atm... Without some serious rewrites, doing that in a general way (for testing this particular problem) would be quite hard I imagine. Internally `runc` does: `files, err := ioutil.ReadDir("/sys/kernel/mm/hugepages")`, aka. reads directly from the system. Since only aarch64(armv8 64bits, not 100% sure about arm32) have hugePages smaller than `1MB`, some serous file mocking, and code rewriting, would be required to test it on other architectures.

Cool, this sounds great! Agree that an integration test is a large undertaking and we shouldn't block this immediate fix on it. Can you please ping me when your upstream runc change is merged?

Thanks for your work here :)

derekwaynecarr commented:
/assign

The feature was tested and used.  As noted consuming hugepages less than 2MB is atypical.  I want to verify this on other OS prior to merging.

derekwaynecarr commented on test/e2e_node/hugepages_test.go:
```diff
@@ -38,19 +37,8 @@ import (
-	// this command takes the expected value and compares it against the actual value for the pod cgroup hugetlb.2MB.limit_in_bytes
-	command := fmt.Sprintf("expected=%v; actual=$(cat /tmp/hugetlb/%v/hugetlb.2MB.limit_in_bytes); if [ \"$expected\" -ne \"$actual\" ]; then exit 1; fi; ", hugePagesLimit.Value(), cgroupFsName)
+// makeHugePagePod returns a pod that requests the the given amount of huge page memory, and execute the given command
+func makeHugePagePod(baseName string, command string, totalHugePageMemory resource.Quantity, hugePageSize resource.Quantity) *apiv1.Pod {
```
I suspect the change made here was to handle environments where 2MB pages were not available?  Can we still ensure that the hugetlb limit files using the cgroupfs are set as expected for the environment that the test is run against?  If I am following this change, we do not actually validate the pod cgroup setting is actually applied.

odinuge replied:
Ish. The tests checks that you can allocate the requested amount, both using the mounted volume, and doing it "manually" in mmap. It also checks that allocation more than the requested amount will fail, and that allocation with another huge page size (same amount of memory) also will crash

derekwaynecarr commented on test/e2e_node/hugepages_test.go:
```diff
@@ -60,22 +48,29 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
 			Volumes: []apiv1.Volume{
 				{
-					Name: "sysfscgroup",
+					Name: "hugetlb",
```
we should still have a test container that mounts the sys/fs/cgroup volume into the pod so we can inspect the actual cgroup settings are applied since the intention is to validate that usage of pre-allocated hugepages have the limit enforced.

derekwaynecarr replied:
if i recall, access to hugepages when consumed by a jvm is not through a volume, so having a test that still read the proper values were written that we expected into the cgroup tree is useful and consistent with our testing for all other cgroup controllers (cpu, memory, pids).

odinuge replied:
Yee, did think about verifying it like the previous test did, so I guess I can keep the old one as well. 

> if i recall, access to hugepages when consumed by a jvm is not through a volume, so having a test that still read the proper values were written that we expected into the cgroup tree is useful and consistent with our testing for all other cgroup controllers (cpu, memory, pids).

Yee. The testing progam tests both via the mounted volum, and via normal mmap with a specified page size(see:
https://github.com/kubernetes/kubernetes/pull/78495/files#diff-e2d845a4981d87782f433ee65956641cR51). (I assume that is the way the jvm does it)

> tree is useful and consistent with our testing for all other cgroup controllers (cpu, memory, pids).

Didn't think about that, keeping the old test then :smile:

odinuge replied:
Readded the old test which was checking the cgroup file now, @derekwaynecarr.

derekwaynecarr commented on test/e2e_node/hugepages_test.go:
```diff
@@ -121,24 +116,39 @@ func configureHugePages() error {
+}
+
+func getTestValues() (hugePageSize resource.Quantity, totalMemory resource.Quantity, pageCount int64) {
+	hugePageSize, _ = getDefaultHugePageSize()
```
changing the test to support a broader set of hugepage configurations is a nice update.  thanks for this.  2MB and 1Gi were most common when this feature was authored on platforms that had desired it.

odinuge replied:
Thanks! I guess 99% will use the default 2MB size, but you will never know...

Another thing outside the scope of the PR: https://github.com/kubernetes/kubernetes/pull/50859/files#diff-f19a7777a36882029db668a427f950feR3517 Is there actually a reason to throw this error? It makes it "impossible" to start kubelet propery when having more than one huge page size with allocated memory. I know that adding support for using multiple page sizes for a pod would require quite a bit work, but each node should be able to support multiple sizes, or am I wrong? Cant find anything about it in the KIP either.. https://github.com/kubernetes/enhancements/tree/master/keps/sig-node

derekwaynecarr replied:
supporting more than one page size per node was requested by @dchen1107 when we first introduced this feature.  in general, i am not aware of how common it is for a workload to demand more than one page size, and nodes that require pre-allocated hugepages are probably purpose built.

derekwaynecarr reviewed (changes requested):
Thank you for finding the bug with kB page sizes.  As noted, most testing had been done on 2MB and 1GI platforms if I recall.  This change is good if we can preserve the testing that actually validates the content expected was written to the pod level cgroup for the host default page size.

odinuge commented on test/utils/image/manifest.go:
```diff
@@ -250,6 +252,7 @@ func initImageConfigs() map[int]Config {
 	configs[VolumeISCSIServer] = Config{e2eRegistry, "volume/iscsi", "2.0"}
 	configs[VolumeGlusterServer] = Config{e2eRegistry, "volume/gluster", "1.0"}
 	configs[VolumeRBDServer] = Config{e2eRegistry, "volume/rbd", "1.0.1"}
+	configs[HugePageTester] = Config{"docker.io", "odinuge/hugepage-tester", "1.0"}
```
This has to be changed to `Config{e2eRegistry, "hugepage-tester", "1.0"}`, and the image has to be built and pushed to the e2eRegistry before merging

odinuge commented:
The fix just got merged into `runc`! I have updated the required dependencies together with runc, and fixed some minor breaking changes. :smile: @derekwaynecarr @mattjmcnaughton

odinuge commented:
Have you had any chance to look at this @mattjmcnaughton & @derekwaynecarr? Would be nice to get this fixed! 😄 

/assign @thockin

mattjmcnaughton commented:
> Have you had any chance to look at this @mattjmcnaughton & @derekwaynecarr? Would be nice to get this fixed! 
> 
> /assign @thockin

Oh wonderful! Apologies that I missed this. I'm going to let @derekwaynecarr take it from here, cause he had some great questions, and I want to be sure he's happy with the answers. Thanks for your work and thanks for the reminder ping :)

/ok-to-test

odinuge commented:
/retest

odinuge commented:
/test pull-kubernetes-e2e-gce-csi-serial

odinuge commented:
/test pull-kubernetes-kubemark-e2e-gce-big

odinuge commented:
/test pull-kubernetes-bazel-test

odinuge commented:
/test pull-kubernetes-bazel-test

odinuge commented:
/retest

Pennyzct commented:
Hi~@odinuge I'm working for ARM and recently about to deploy kata containers with k8s on AArch64. 
thanks for your fix!!!! After adding you patch, kubernete 1.14.x could be finally working. ;)
I'm really looking forward to seeing it merged😍, and cherry picked back to v1.14.x. ;)

odinuge commented:
Yeah @Pennyzct, would be nice getting this merged as soon as possible, and cherry picking it to `1.14.X` and `1.15.X` (will not be a part of the `1.15` release), so that we finally can get AArch64 working again!

CC: @dims, Ref. https://github.com/kubernetes/kubernetes/pull/78187 i guess you may want to see this PR :smile:

odinuge committed:
Update dependency opencontainer/runc

odinuge committed:
Update dependency sirupsen/logrus

odinuge commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

odinuge commented:
/test pull-kubernetes-bazel-test

odinuge commented:
/test pull-kubernetes-e2e-gce

derekwaynecarr commented:
the change is fine, it looks like we will keep the existing tests with 2MB page testing which is fine.

/lgtm
/approve

odinuge commented:
/cc @dims

odinuge commented:
/assign @liggitt @thockin

(for root approval for go.* file)

odinuge commented:
Since https://github.com/kubernetes/kubernetes/pull/79287, the bump of libseccomp-go, is now in tide and about to get merged, would be nice if we if we could get this in as well, since `runc` is the dependency that depends on `libseccomp-go`! @liggitt

liggitt commented on go.mod:
```diff
@@ -111,7 +112,7 @@ require (
 	github.com/opencontainers/go-digest v1.0.0-rc1
 	github.com/opencontainers/image-spec v1.0.1 // indirect
-	github.com/opencontainers/runc v0.0.0-20181113202123-f000fe11ece1
+	github.com/opencontainers/runc v1.0.0-rc2.0.20190611121236-6cc515888830
```
how was this revision selected? can we pin to a tagged release?

odinuge replied:
It is the first commit including the required fix in runc. https://github.com/opencontainers/runc/pull/2065. They have no release containing the new fix.

odinuge replied:
I can rebase to the newest commit f4982d86f7fde0b6f953cc62ccc4022c519a10a9, but there has been no code changes since `6cc515888830`

liggitt commented:
I'll defer to node reviewers on whether the runc bump is required for this, but if the supported suffixes are fixed, would it be more straightforward to update our internal list of supported suffixes?

odinuge commented:
> would it be more straightforward to update our internal list of supported suffixes?

That _is_ possible, but without the fix in runc, the code _will still_ crash on arm64 because the `runc`code using `kB` will be executed here: https://github.com/kubernetes/kubernetes/blob/5f88b0d18747d257053ae04cafb096a01b92d41a/pkg/kubelet/cm/cgroup_manager_linux.go#L399 😢 

So no, that will not work. :/

odinuge commented:
Also, imo. it better to keep using the list from `runc`, with their corresponding tests, to avoid a regression. Or, what do you think @liggitt 🤔?

liggitt commented:
> Also, imo. it better to keep using the list from `runc`, with their corresponding tests, to avoid a regression. Or, what do you think @liggitt 🤔?

If the list is just a representation of what is independently supported by cgroups, then I don't see a lot of value in using a constant from another project.

odinuge commented:
Hi @liggitt 😄 

TL;DR: We still need to upgrade `runc`. If using a list with prefixes for huge page sizes inside kubernetes still is preferred, I will be happy to switch to that.

> If the list is just a representation of what is independently supported by cgroups, then I don't see a lot of value in using a constant from another project.

Ok, I can definitely do that. *But*, the problem was just that a huge list of project copied the list (the wrong list with `kB`, and not `KB`) from the same project (i guess it comes from runc), making the bug hard to fix, since it is hardcoded in a lot of places. We would still implicitly depend on the `runc` list quite a lot in kubernetes, so an updated version of runc is required either way.

We could in theory copy the code used to get a list of the supported sizes, the var`HugePageSize` we use from runc, but that would just result in doing the same work twice. 

The code in `runc` is also required here:

https://github.com/kubernetes/kubernetes/blob/5f88b0d18747d257053ae04cafb096a01b92d41a/pkg/kubelet/cm/cgroup_manager_linux.go#L570

So with the current version of runc, and switching from `kB`to `KB` in kubernetes, will make `GetResourceStats` return `nil` as stats, together with an error saying that something is wrong with cgroups.

All this currently only affects arm64 (making kubelet fail), but it it still a fact that our cgroup code is wrong. :/

cc @derekwaynecarr

liggitt commented:
> Ok, I can definitely do that. _But_, the problem was just that a huge list of project copied the list (the wrong list with `kB`, and not `KB`) from the same project (i guess it comes from runc), making the bug hard to fix, since it is hardcoded in a lot of places. We would still implicitly depend on the `runc` list quite a lot in kubernetes, so an updated version of runc is required either way.

I'm assuming we'll want to pick the kB -> KB fix back to previous releases. Doing so by just changing our inline list will make this *much* easier than also needing to bump dependencies, right?

liggitt commented:
> I'm assuming we'll want to pick the kB -> KB fix back to previous releases. Doing so by just changing our inline list will make this _much_ easier than also needing to bump dependencies, right?

re-read your comment, sounds like the runc bump is also required to fix the issue. If that's true, then using the constant from runc is more reasonable, but makes fixing this in past releases harder.

odinuge commented:
Yes, correct, the `runc` bump **is** required. I think this should be cherry-picked into the previous releases, and as you say, I guess it would require some work (especially pre `go.mod`).

odinuge commented:
Did a rebase now. As you see above @liggitt, @derekwaynecarr (from sig-node & approver in `pkg/kubelet`) commented with both `/lgtm` and `/approve` (the lgtm was removed because of the rebase), so the code has already been approved from someone in sig-node. But because of the changes in go.mod & go.sum, that is not enough.

Also, how far back do you think we should cherry-pick this?

liggitt commented:
/lgtm
/approve

liggitt commented:
I'll defer to sig-node on the severity of the issue this is fixing vs the risk of back-porting

odinuge commented:
That sounds like a good idea @liggitt! Thanks! :smile:

Pennyzct commented:
Hi~@odinuge @liggit again~ Penny from ARM. super excited to see this PR got merged. ;)
Looking forward to see this patch backed port to v1.14 and v1.15. Otherwise, if you try to set up k8s cluster on AArch64 with kernel 5.0+, building from the scratch maybe the only option.😭😢

odinuge commented:
Hi @Pennyzct! Yes, currently, the only way to run k8s on AArch64 is to compile the master branch of k8s. I think this can become a huge pain point when people gets their hands on the new Raspberry Pi, and starts testing with 5.0 kernels and k8s. :/

 I will make a patch for the `v1.15` branch as quick as I can, and since the difference between master and v1.15 isn't that big, it shouldn't be a huge problem. When it comes to back porting it further, to v1.14 etc., I need some input from the sig-node team to hear if they want it or not, ref.

> I'll defer to sig-node on the severity of the issue this is fixing vs the risk of back-porting

I can certainly do all the back porting, as long as there is will in sig-node to get it merged! 😄 

Any thoughts @derekwaynecarr & @mattjmcnaughton?

mattjmcnaughton commented:
I agree let's start with 1.15.

For me, the largest question is how often someone would be using this `hugetlb` functionality, and in what scenarios. I'd be hesitant to upgrade runc in past versions if the `hugetlb` functionality either didn't enjoy widespread use or was predominantly used in non-production instances (in which an upgrade to 1.15.x would be less onerous).

Thoughts?

derekwaynecarr commented:
The issue is how often hugetlb is used with that specific page size.  I will look at the backport, I am not adverse to it.
